File: frontend/src/types/schedule.ts
import type { CapacityResourceType } from "../lib/scheduleConfig";

export interface Occupant {
  booking_id: string;
  patient_name?: string | null;
  pathway_name?: string | null;
}

export interface ResourceSlot {
  slot_index: number;
  occupied: number;
  capacity: number;
  blocked: boolean;
  free: boolean;
  occupants: Occupant[];
  /** Patient-column uptake/gap segment — resources free, patient still present. */
  is_uptake?: boolean;
}

export interface ResourceColumn {
  resource_id: string | null;
  resource_type: string;
  name: string | null;
  capacity: number;
  slots: ResourceSlot[];
}

export interface ScheduleDay {
  date: string;
  day_start_hour: number;
  day_end_hour: number;
  slot_minutes: number;
  slots_per_day: number;
  columns: ResourceColumn[];
  /** IANA zone the clinic operates in — slot indices are offsets in this zone. */
  clinic_timezone: string;
  /** Server's current time, ISO-8601 with offset. The authority for the now-line. */
  clinic_now: string;
  clinic_today: string;
  current_slot_index: number;
}

export interface SlotPatchRequest {
  date: string;
  resource_type: CapacityResourceType;
  slot_indices: number[];
  blocked: boolean;
}

File: backend/app/schemas/schedule.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.resource import ResourceType


class SlotPatchRequest(BaseModel):
    date: date
    resource_type: ResourceType
    slot_indices: list[int] = Field(min_length=1)
    blocked: bool


class OccupantOut(BaseModel):
    booking_id: UUID
    patient_name: str | None = None
    pathway_name: str | None = None


class ResourceSlotOut(BaseModel):
    slot_index: int
    occupied: int
    capacity: int
    blocked: bool = False
    free: bool
    occupants: list[OccupantOut] = []
    # True for gap/uptake cells (Patient column gaps for admin; GAP column for patients).
    is_uptake: bool = False


class ResourceColumnOut(BaseModel):
    resource_id: UUID | None
    resource_type: str
    name: str | None = None
    capacity: int
    slots: list[ResourceSlotOut]


class ScheduleDayOut(BaseModel):
    date: date
    day_start_hour: int
    day_end_hour: int
    slot_minutes: int
    slots_per_day: int
    columns: list[ResourceColumnOut]
    # The backend owns "now". Clients render the now-line from these instead of
    # their own clock, so a patient in another timezone sees the clinic's day.
    clinic_timezone: str
    clinic_now: datetime
    clinic_today: date
    current_slot_index: int
